posthog/clickhouse/property_groups.py

```diff
@@ -1,40 +1,76 @@
+    ) -> None:
         self.__cluster = cluster
-        self.__table = table
-        self.__source_column = source_column
-        self.__groups: MutableMapping[str, PropertyGroupDefinition] = {}
+        self.__groups = groups
+
+    def __get_map_column_name(self, column: ColumnName, group_name: PropertyGroupName) -> str:
+        return f"{column}_group_{group_name}"
 
-    def register(self, name: str, definition: PropertyGroupDefinition) -> None:
-        assert name not in self.__groups, "property group names can only be used once"
-        self.__groups[name] = definition
+    def __get_column_definition(self, table: TableName, column: ColumnName, group_name: PropertyGroupName) -> str:
+        group_definition = self.__groups[table][column][group_name]
+        map_column_name = self.__get_map_column_name(column, group_name)
+        column_definition = f"{map_column_name} Map(String, String)"
+        if not group_definition.is_materialized:
+            return column_definition
+        else:
+            return f"""\
+                {column_definition}
+                MATERIALIZED mapSort(
+                    mapFilter((key, _) -> {group_definition.key_filter_expression},
+                    CAST(JSONExtractKeysAndValues({column}, 'String'), 'Map(String, String)'))
+                )
+                CODEC({group_definition.codec})
+            """
 
-    def __get_map_expression(self, definition: PropertyGroupDefinition) -> str:
-        return f"mapSort(mapFilter((key, _) -> {definition.filter_expression}, CAST(JSONExtractKeysAndValues({self.__source_column}, 'String'), 'Map(String, String)')))"
+    def __get_index_definitions(
+        self, table: TableName, column: ColumnName, group_name: PropertyGroupName
+    ) -> Iterable[str]:
+        group_definition = self.__groups[table][column][group_name]
+        if not group_definition.is_materialized:
+            return
 
-    def get_alter_create_statements(self, name: str) -> Iterable[str]:
-        column_name = f"{self.__source_column}_group_{name}"
-        definition = self.__groups[name]
-        return [
-            f"ALTER TABLE {self.__table} ON CLUSTER {self.__cluster} ADD COLUMN {column_name} Map(String, String) MATERIALIZED {self.__get_map_expression(definition)} CODEC({definition.codec})",
-            f"ALTER TABLE {self.__table} ON CLUSTER {self.__cluster} ADD INDEX {column_name}_keys_bf mapKeys({column_name}) TYPE bloom_filter",
-            f"ALTER TABLE {self.__table} ON CLUSTER {self.__cluster} ADD INDEX {column_name}_values_bf mapValues({column_name}) TYPE bloom_filter",
-        ]
+        map_column_name = self.__get_map_column_name(column, group_name)
+        yield f"{map_column_name}_keys_bf mapKeys({map_column_name}) TYPE bloom_filter"
+        yield f"{map_column_name}_values_bf mapValues({map_column_name}) TYPE bloom_filter"
 
+    def get_create_table_pieces(self, table: TableName) -> Iterable[str]:
```

Comment:
Not sure what the correct terminology is here. They're not expressions, but also not whole statements, so just went with something generic.